solr/core/src/java/org/apache/solr/core/SolrCore.java

```diff
@@ -491,22 +490,27 @@ public long getIndexSize() {
   }
 
   public long getCachedIndexSize() {
-    if (indexSize.get() == null) {
-      if (log.isDebugEnabled()) {
-        log.debug("Recalculating index size for {}", getName());
+    SolrRequestInfo requestInfo = SolrRequestInfo.getRequestInfo();
+    if (requestInfo != null) {
+      Map<Object,Object> contextMap = requestInfo.getReq().getContext();
+      Long cachedIndexSize = (Long)contextMap.get(cachedIndexSizeKeyName(getName()));
+      if (cachedIndexSize == null) {
+        if (log.isDebugEnabled()) {
+          log.debug("Recalculating index size for {}", getName());
+        }
+        cachedIndexSize = getIndexSize();
+        contextMap.put(cachedIndexSizeKeyName(getName()), cachedIndexSize);
+      } else if (log.isDebugEnabled()) {
+        log.debug("reusing previous index size for {}", getName());
       }
-      indexSize.set(getIndexSize());
-    } else if (log.isDebugEnabled()) {
-      log.debug("reusing previous index size for {}", getName());
+      return cachedIndexSize;
+    } else {
+      return getIndexSize();
     }
-    return indexSize.get();
   }
 
-  public void clearIndexSizeCache() {
-    if (log.isDebugEnabled()) {
-      log.debug("Clearing index size cache for {}", getName());
-    }
-    indexSize.remove();
+  private String cachedIndexSizeKeyName(String name) {
```

Comment:
I was thinking maybe we'd need another map inside the context to key by core name but your approach here is simpler and adequate.  Please add a comment on why we're doing this -- it may not be apparent that we're avoiding collisions across cores for the same metrics request.

Reply:
Yes, adding a new map is nicer but it's probably an overkill here. I'll add a comment, that will help.